genienlp/data_utils/bootleg.py

```diff
@@ -33,108 +33,21 @@
-    line = {}
-    line['sentence'] = getattr(ex, task.utterance_field)
-
-    assert len(label) == 7
-    line['aliases'], line['spans'], line['cands'] = label['aliases'], label['spans'], label['cands']
-    line['cand_probs'] = list(map(lambda item: list(item), label['cand_probs']))
-    tokens_type_ids, tokens_type_probs = bootleg_annotator.bootleg.collect_features_per_line(
-        line, args.bootleg_prob_threshold, getattr(task, 'TTtype2qid', None)
-    )
-
-    if task.utterance_field == 'question':
-        for i in range(len(tokens_type_ids)):
-            ex.question_feature[i].type_id = tokens_type_ids[i]
-            ex.question_feature[i].type_prob = tokens_type_probs[i]
-
-    else:
-        for i in range(len(tokens_type_ids)):
-            ex.context_feature[i].type_id = tokens_type_ids[i]
-            ex.context_feature[i].type_prob = tokens_type_probs[i]
-
-    ex.question_plus_types = task.add_type_tokens(ex.question, ex.question_feature, args.add_types_to_text)
-    ex.context_plus_types = task.add_type_tokens(ex.context, ex.context_feature, args.add_types_to_text)
-
-    return ex
-
-
-def extract_features_with_annotator(examples, bootleg_annotator, args, task):
-    with torch.no_grad():
-        bootleg_inputs = []
-        for ex in examples:
-            bootleg_inputs.append(getattr(ex, task.utterance_field))
-
-        bootleg_labels = bootleg_annotator.label_mentions(bootleg_inputs)
-
-        keys = tuple(bootleg_labels.keys())
-        values = list(bootleg_labels.values())
-        values_unpacked = list(zip(*values))
-
-        bootleg_labels_unpacked = [dict(zip(keys, values)) for values in values_unpacked]
-
-        for i in range(len(examples)):
-            ex = examples[i]
-            label = bootleg_labels_unpacked[i]
-            examples[i] = bootleg_process_examples(ex, bootleg_annotator, args, label, task)
-
-
-def init_bootleg_annotator(args, device, bootleg=None):
-    # instantiate a bootleg object to load config and relevant databases
-    if bootleg is None:
-        bootleg = Bootleg(args)
-    bootleg_config = bootleg.create_config(bootleg.fixed_overrides)
-
-    # instantiate the annotator class. we use annotator only in server mode
-    # for training we use bootleg functions which preprocess and cache data using multiprocessing, and batching to speed up NED
-    bootleg_annotator = BootlegAnnotator(
-        config=bootleg_config,
-        device='cpu' if device.type == 'cpu' else 'cuda',
-        min_alias_len=args.min_entity_len,
-        max_alias_len=args.max_entity_len,
-        cand_map=bootleg.cand_map,
-        threshold=args.bootleg_prob_threshold,
-        model_name=args.bootleg_model,
-        verbose=False,
-    )
-    # collect all outputs now; we will filter later
-    bootleg_annotator.set_threshold(0.0)
-    setattr(bootleg_annotator, 'bootleg', bootleg)
-    return bootleg_annotator
-
-
 class Bootleg(object):
```

Comment:
So now we have "Bootleg", "BootlegAnnotator", and "BootlegAnnotator renamed to BootlegClass". This is beyond confusing. Can we pick better names, or add doc comments to distinguish?

Reply:
Our `BootlegAnnotator` class is a wrapper for bootleg's `BootlegAnnotator` class. I can rename theirs to annotator in the import and add a docstring for our `Bootleg` and `BootlegAnnotator` classes.

Reply:
What's the purpose of our BootlegAnnotator, and what's the purpose of the Bootleg class? Maybe we can find three distinct non-confusing names.

Reply:
`Bootleg` class is a wrapper for all the functionalities we need from bootleg. It calls the appropriate functions from bootleg, processes the collected results, extract the needed features, and add those features to our examples.

`BootlegAnnotator` does the same but in a on-the-fly fashion useful for server mode when input data is unknown ahead of time.

I kinda like the current names but if the conflict is with the original `BootlegAnnotator` we can rename ours to `BootlegAnnotatorWrapper`.